crypto/sm2/sm2_pmeth.c

```diff
@@ -167,6 +197,14 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
         EC_GROUP_set_asn1_flag(dctx->gen_group, p1);
         return 1;
 
+    case EVP_PKEY_CTRL_SM2_SET_UID:
+        dctx->uid = OPENSSL_strdup(p2);
+        return 1;
+
+    case EVP_PKEY_CTRL_SM2_GET_UID:
+        *(char**)p2 = OPENSSL_strdup(dctx->uid);
```

Comment:
Do you mean "is it really true that SM2 UIDs will always be null terminated strings"? I believe this is the case. But perhaps @ronaldtse knows more.

Reply:
I meant do we need to create a copy of it for the getter? Can't we just return a direct reference to `dctx->uid` (i.e. in OpenSSL terms we treat this as a `get0` instead of a `get1`)

Reply:
@randombit SM2 UIDs (in DSA, PKE) makes no assumption to NULL-termination, the length of the UID is explicitly coded in `ENTL`. Perhaps there's no need to make a copy as suggested by @mattcaswell ?